storage_storer/metadata_storer.go

```diff
@@ -0,0 +1,67 @@
+package storage_storer
+
+type MetadataStorer struct {
+	Locker MultiLock
+	Storer metadataStorerImplementation
+}
+
+func (s *MetadataStorer) Create(key string, value string) error {
+	lock, err := s.Locker.Lock(key)
+	if err != nil {
+		return err
+	}
+	defer s.Locker.Unlock(lock)
+
+	return s.Storer.Set(key, value)
+}
+
+func (s *MetadataStorer) Update(key string, updateFn func(string) (string, error)) error {
+	lock, err := s.Locker.Lock(key)
+	if err != nil {
+		return err
+	}
+	defer s.Locker.Unlock(lock)
+
+	currentValue, err := s.Storer.Get(key)
+	if err != nil {
+		return err
+	}
+
+	newValue, err := updateFn(currentValue)
+	if err != nil {
+		return err
+	}
+
+	return s.Storer.Set(key, newValue)
+}
+
+func (s *MetadataStorer) List(prefix string) (map[string]string, error) {
+	// TODO: how do we lock a prefix?
```

Comment:
good question

Reply:
thinking out loud, we could create another map to track that state, but that feels like a lot of boiler plate.

Reply:
Could also make the locker interface prefix aware, or just lock the entire thing when listing